gcode.c

```diff
@@ -599,14 +607,16 @@ char *gc_normalize_block (char *block, status_code_t *status, char **message)
                         if(message && *message == NULL) {
 #if NGC_EXPRESSIONS_ENABLE
```

Comment:
@terjeio Should handling of `(DEBUG,` and `(PRINT,` be moved to `ngc_flowctrl`'s `onGcodeComment`? 🤔

Reply:
No. Since substitution is done in _ngc_expr.c_ it belongs there, and also because the` grbl.on_gcode_comment()` event is not able to return a string (and should not be changed to accomodate that).

Your `grbl.on_string_substitution()` event could be set up at startup to point to a new function in _ngc_expr.c_, making it overridable. And then your string register handling could (or rather should?) be moved to a plugin, in [this repo](https://github.com/grblHAL/Plugins_misc/)? `grbl.on_string_substitution`  has to be changed to a more descriptive name if doing so...

Reply:
Aha, maybe I should make a separate PR with just the string substitution-stuff, and move string registers to a plugin-repo? 🤔

Reply:
>            case ')':
                if(comment && !gc_state.skip_blocks) {
                    *s1 = '\0';
                    if(!hal.driver_cap.no_gcode_message_handling) {

                        if(message && *message == NULL) {

                            if(grbl.on_process_gcode_comment)
                                *message = grbl.on_process_gcode_comment(comment);

                            if(*message == NULL) {
                                size_t len = s1 - comment - 3;

                                if(!strncasecmp(comment, "MSG,", 4) && (*message = malloc(len))) {

                                    comment += 4;
                                    while(*comment == ' ') {
                                        comment++;
                                        len--;
                                    }
                                    memcpy(*message, comment, len);
                                }
                            }
                        }
                    }
                    if(*comment && *message == NULL && grbl.on_gcode_comment)
                        *status = grbl.on_gcode_comment(comment);
                }
                comment = NULL;
                break

Reply:
The previous comment was somehow posted before I finished, trying again:

> Aha, maybe I should make a separate PR with just the string substitution-stuff, and move string registers to a plugin-repo?

I'll update the core, you add a plugin for string registers _and_ string substitution that includes string registers. This since I'll make comment handling completely overridable even for [MSG, ...), in gcode.c:

```
            case ')':
                if(comment && !gc_state.skip_blocks) {
                    *s1 = '\0';
                    if(!hal.driver_cap.no_gcode_message_handling) {

                        if(message && *message == NULL) {

                            if(grbl.on_process_gcode_comment)
                                *message = grbl.on_process_gcode_comment(comment);

                            if(*message == NULL) {
                                size_t len = s1 - comment - 3;

                                if(!strncasecmp(comment, "MSG,", 4) && (*message = malloc(len))) {

                                    comment += 4;
                                    while(*comment == ' ') {
                                        comment++;
                                        len--;
                                    }
                                    memcpy(*message, comment, len);
                                }
                            }
                        }
                    }
                    if(*comment && *message == NULL && grbl.on_gcode_comment)
                        *status = grbl.on_gcode_comment(comment);
                }
                comment = NULL;
                break;
```

`grbl.on_process_gcode_comment` will be defaulted to handle (PRINT, ..) and (DEBUG, ...) when expressions is enabled.
You can choose to do string replacements only and use the default handler for numeric ones, or replace the handler completely. Note that the `grbl.on_process_gcode_comment` signature does not allow for a status code return - this is deliberate since message comments should never fail with an error.

You may choose where to publish a string register plugin, you can add it to your own repo or make a PR for the [misc plugins repo](https://github.com/grblHAL/Plugins_misc/). The Web Builder can build with repos located outside grblHAL, but local builds cannot without manually adding your code to the source tree - at least not without adding it as a submodule which I am reluctant to do.

FYI I have already added string _parameter_ support to ngc_param.c (not yet comitted) in order to handle named o calls, you may build string registers on top of this or add it to your plugin. It is different from your string register code in that it does not return status codes and it will remove a string parameter on failure to allocate for it. Also, it will not reallocate on changes if the new length is less than the previous, this as an attempt to reduce heap fragmentation.